jobqueue/src/main/java/com/birbit/android/jobqueue/JobManagerThread.java

```diff
@@ -123,7 +136,35 @@ private void handleAddJob(AddJobMessage message) {
             JqLog.e(t, "job's onAdded did throw an exception, ignoring...");
         }
         callbackManager.notifyOnAdded(jobHolder.getJob());
-        consumerManager.onJobAdded();
+        if (insert) {
+            consumerManager.onJobAdded();
+        } else {
+            cancelSafely(jobHolder, CancelReason.SINGLE_INSTANCE_ID_QUEUED);
+            callbackManager.notifyOnDone(jobHolder.getJob());
+        }
+    }
+
+    /**
+     * Returns a queued job with the same single id. If any matching non-running job is found,
+     * that one is returned. Otherwise any matching running job will be returned.
+     */
+    private JobHolder findJobBySingleId(/*Nullable*/String singleIdTag) {
+        if (singleIdTag != null) {
+            queryConstraint.clear();
+            queryConstraint.setTags(new String[]{singleIdTag});
+            queryConstraint.setTagConstraint(TagConstraint.ANY);
+            Set<JobHolder> jobs = nonPersistentJobQueue.findJobs(queryConstraint);
```

Comment:
we can directly ask the consumer manager for running jobs first so that we can avoid calling the queues.